[5.1] [FileManager] Revert r347205 to avoid PCH file-descriptor leak.

jrose-apple opened the pull request:
Cherry-pick of https://reviews.llvm.org/D57165 (from January!) to fix a file descriptor leak. See there for details.

rdar://problem/49785385

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
**Explanation**: A change in open-source LLVM to fix a bug where opening file descriptors *wouldn't* be cached ended up creating a bug where file descriptors wouldn't be *closed,* which is much worse. This change reverts that original fix, and has been in open-source LLVM since January.

**Scope**: Affects many uses of Clang's `FileManager`.

**Risk**: Low: public LLVM has been living on this, and the revert brings us back to the behavior in Swift 5.0.

**Testing**: The revert includes a new regression test, and I also verified that this fixes the issue with a precompiled bridging header in rdar://problem/49785385.

**Reviewer**: Originally reviewed by Nico Weber.

jrose-apple commented:
@swift-ci Please test

shahmishal commented:
@swift-ci test

vsapsai commented on lib/Basic/FileManager.cpp:
```diff
@@ -295,11 +279,13 @@ const FileEntry *FileManager::getFile(StringRef Filename, bool openFile,
   UFE.Dir     = DirInfo;
   UFE.UID     = NextFileUID++;
   UFE.UniqueID = Status.getUniqueID();
-  UFE.IsNamedPipe = Status.getType() == llvm::sys::fs::file_type::fifo_file;
```
I think you need to keep this line, though I haven't tested it. Last time it was changed in [a different commit](https://github.com/apple/swift-clang/commit/e6138c2236eeb616c2f2b9774be0640fb0e026db#diff-0025af005307891b5429b6a834823d5eR295), not the one you are reverting.

jrose-apple replied:
Oops. Bad merge resolution, thank you for catching this!

vsapsai reviewed (changes requested):
`UFE.IsNamedPipe` change (commented inline) looks suspicious, the rest looks good.

jrose-apple committed:
[FileManager] Revert r347205 to avoid PCH file-descriptor leak.

Summary:
r347205 fixed a bug in FileManager: first calling
getFile(shouldOpen=false) and then getFile(shouldOpen=true) results in
the file not being open.

Unfortunately, some code was (inadvertently?) relying on this bug: when
building with a PCH, the file entries are obtained first by passing
shouldOpen=false, and then later shouldOpen=true, without any intention
of reading them. After r347205, they do get unneccesarily opened.
Aside from extra operations, this means they need to be closed. Normally
files are closed when their contents are read. As these files are never
read, they stay open until clang exits. On platforms with a low
open-files limit (e.g. Mac), this can lead to spurious file-not-found
errors when building large projects with PCH enabled, e.g.
  https://bugs.chromium.org/p/chromium/issues/detail?id=924225

Fixing the callsites to pass shouldOpen=false when the file won't be
read is not quite trivial (that info isn't available at the direct
callsite), and passing shouldOpen=false is a performance regression (it
results in open+fstat pairs being replaced by stat+open).

So an ideal fix is going to be a little risky and we need some fix soon
(especially for the llvm 8 branch).
The problem addressed by r347205 is rare and has only been observed in
clangd. It was present in llvm-7, so we can live with it for now.

Reviewers: bkramer, thakis

Subscribers: ilya-biryukov, ioeric, kadircet, cfe-commits

Differential Revision: https://reviews.llvm.org/D57165

git-svn-id: https://llvm.org/svn/llvm-project/cfe/trunk@352079 91177308-0d34-0410-b5e6-96231b3b80d8

jrose-apple commented:
@swift-ci test
